src/erasure-code/consistency/ECEncoder.cc

```diff
@@ -0,0 +1,207 @@
+#include "ECEncoder.h"
+#include <typeinfo>
+#include "common/errno.h"
+#include "osd/ECUtil.h"
+#include "osd/ECUtilL.h"
+
+template class ceph::consistency::ECEncoder<stripe_info_l_t>;
+template class ceph::consistency::ECEncoder<ECUtil::stripe_info_t>;
+
+using stripe_info_l_t = ECLegacy::ECUtilL::stripe_info_t;
+
+template <typename SInfo>
+ceph::consistency::ECEncoder<SInfo>::ECEncoder(ceph::ErasureCodeProfile profile, int chunk_size) :
+profile(profile),
+chunk_size(chunk_size)
+{
+  int r = ec_init_plugin(&ec_impl);
+  if (r < 0) {
+    std::cerr << "Failed to initialize plugin: " << r << std::endl;
+  }
+  ec_init_sinfo(&ec_impl, &stripe_info);
+}
+
+/**
+ * Initialize the ErasureCodeInterfaceRef needed to perform encode.
+ *
+ * @param ec_impl Pointer to plugin being initialized
+ * @returns int 0 if successful, otherwise 1
+ */
+template <typename SInfo>
+int ceph::consistency::ECEncoder<SInfo>::ec_init_plugin(ceph::ErasureCodeInterfaceRef *ec_impl)
+{
+  auto plugin = profile.find("plugin");
+  if (plugin == profile.end()) {
+    std::cerr << "Invalid profile: plugin not specified." << std::endl;
+    return 1;
+  }
+
+  std::stringstream ss;
+  std::string dir = g_conf().get_val<std::string>("erasure_code_dir");
+  ceph::ErasureCodePluginRegistry::instance().factory(plugin->second,
+                                                      dir, profile,
+                                                      ec_impl, &ss);
+  if (!*ec_impl) {
+    std::cerr << "Invalid profile: " << ss.str() << std::endl;
+    return 1;
+  }
+
+  return 0;
+}
+
+/**
+ * Initialize the stripe_info_t needed to perform encode. Optimized version for new EC.
+ *
+ * @param ec_impl Pointer to plugin object
+ * @param s Pointer to the stripe info object associated with the EC profile
+ * @returns int 0 if successful, otherwise 1
+ */
+template <typename SInfo>
+int ceph::consistency::ECEncoder<SInfo>::ec_init_sinfo(ceph::ErasureCodeInterfaceRef *ec_impl,
```

Comment:
Passing a uniqu_p as an in/out parameter is strange. And having an in/out parameter when you only need an out, is not the best option.
why not use the following signature:
std::unique_ptr<ECUtil::stripe_info_t> consistency::ECEncoder<SInfo>::ec_init_sinfo(
  const ceph::ErasureCodeInterfaceRef& ec_impl)
?